torch/testing/_asserts.py

```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
-            return exc, None
+    exc, actual = _to_tensor(actual)
+    if exc:
+        return exc, None
 
-        result = cast(_ParsedInputs, result)
-        actual_lst.append(cast(Tensor, result.actual))
-        expected_lst.append(cast(Tensor, result.expected))
+    exc, expected = _to_tensor(expected)
+    if exc:
+        return exc, None
 
-    return None, _ParsedInputs(actual_lst, expected_lst)
+    return None, _TensorPair(actual, expected)
 
 
-def _parse_mappings(actual: Mapping, expected: Mapping) -> Tuple[Optional[Exception], Optional[_ParsedInputs]]:
-    """Parses sequences of scalar-or-array-like pairs.
+def _parse_inputs(actual: Any, expected: Any) -> Tuple[Optional[Exception], Optional[Union[_TensorPair, List, Dict]]]:
+    """Parses the positional inputs by constructing :class:`_TensorPairs` from corresponding array-or-scalar-likes.
 
-    Regardless of the input types, the sequences are returned as :class:`dict`.
+    :class:`~collections.abc.Sequence`'s or :class:`~collections.abc.Mapping`'s are parsed elementwise. Parsing is
+    performed recursively and thus nested containers are supported. The hierarchy of the containers is preserved, but
+    sequences are returned as :class:`list` and mappings as :class:`dict`.
 
     Args:
-        actual: Actual mapping array-or-scalar-likes.
-        expected: Expected mapping array-or-scalar-likes.
+        actual (Any): Actual input.
+        expected (Any): Expected input.
 
     Returns:
-        (Optional[Exception], Optional[_ParsedInputs]): The two elements are orthogonal, i.e. if the first ``is None``
-            the second will not and vice versa. Returns a :class:`AssertionError` if the keys of :attr:`actual` and
-            :attr:`expected` do not match. Additionally, returns any exception from
-            :func:`_parse_array_or_scalar_like_pair`.
+        (Tuple[Optional[Exception], Optional[Union[_TensorPair, List, Dict]]]): The two elements are orthogonal, i.e.
+            if the first ``is None`` the second will be valid and vice versa. Returns an :class:`AssertionError` if the
+            length of two sequences or the keys of two mappings do not match. Additionally, returns any exception from
+            :func:`_to_tensor_pair`.
     """
     exc: Optional[Exception]
 
-    actual_keys = set(actual.keys())
-    expected_keys = set(expected.keys())
-    if actual_keys != expected_keys:
-        missing_keys = expected_keys - actual_keys
-        additional_keys = actual_keys - expected_keys
-        exc = AssertionError(
-            f"The keys of the mappings do not match:\n\n"
-            f"Missing keys in the actual mapping: {sorted(missing_keys)}\n"
-            f"Additional keys in the actual mapping: {sorted(additional_keys)}\n"
-        )
-        return exc, None
+    # We explicitly exclude str's here since they are self-referential and would cause an infinite recursion loop:
+    # "a" == "a"[0][0]...
```

Comment:
Of course the same is true for other self-referential sequences. Safeguarding this for the general case is more involved and should only be implemented if there is an actual use case.